editor/code_editor.cpp

```diff
@@ -1129,6 +1129,53 @@ void CodeTextEditor::trim_trailing_whitespace() {
 	}
 }
 
+void CodeTextEditor::trim_final_newlines() {
+	int final_line = text_editor->get_line_count() - 1;
+	int check_line = final_line;
+
+	String line = text_editor->get_line(final_line);
+
+	while (line.length() == 0 && check_line > -1) {
+		line = text_editor->get_line(--check_line);
+	}
+
+	if (check_line++ < final_line) {
+		text_editor->begin_complex_operation();
+
+		text_editor->remove_text(check_line, 0, final_line, 0);
+
+		Vector<int> caret_edit_order = text_editor->get_caret_index_edit_order();
+		for (int i = 0; i < caret_edit_order.size(); i++) {
+			int c = caret_edit_order[i];
+
+			if (text_editor->has_selection(c)) {
+				int from_line = text_editor->get_selection_from_line(c);
+				int to_line = text_editor->get_selection_to_line(c);
+
+				if (from_line >= check_line) {
+					text_editor->deselect(c);
+				} else if (to_line > check_line) {
+					int from_column = text_editor->get_selection_from_column(c);
+
+					text_editor->select(from_line, from_column, check_line, 0, c);
+				}
+			}
+
+			int loc = text_editor->get_caret_line(c);
+			if (loc > check_line) {
+				if (i == 0) {
+					text_editor->set_caret_line(check_line);
+				} else {
+					text_editor->remove_caret(c);
+				}
+			}
+		}
```

Comment:
This is my first pass at implementing the caret and selection handling for when either are outside of the area that is now going to be deleted. 

I have gotten it down to only a small fraction of my manual tests resulting in errors, but would love to know if this is heading in the right direction in case I am completely off base. 

Known passing:
- Caret inside section to be deleted
- Caret outside section to be deleted
- Selection inside section to be deleted
- Selection outside section to be deleted
- Selection overlapping section to be deleted when caret is at the bottom of selection

Known failing: 
- Selection overlapping section to be deleted when caret is at the top of selection
  - Note: This failure occurs when the mouse clicks somewhere else, exiting the selection mode. It appears that something is still referencing the previous to line despite calling `select` for the caret. 
- Multiple carets inside section to be deleted
  - Note: There is one error that the caret index is out of range in `get_caret_line` and then one error per caret for `get_line_wrap_index_at_column` with line out of range.
  
Any suggestions on this would be appreciated.

Reply:
> Multiple carets inside section to be deleted

Fixed by iterating backwards over the cursors. Previously I was deleting then trying to access.

Reply:
No longer needed due to @kitbdev's comment + PR.